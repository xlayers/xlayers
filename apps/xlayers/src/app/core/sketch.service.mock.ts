import { SketchMSData } from '@xlayers/sketchtypes';
import { getFlatLayerMock } from '../editor/preview/viewer/layer/layer.component.mock';

export function getSketchDataMock() {
  return {
    previews: [{}],
    pages: [getFlatLayerMock(9)],
  } as SketchMSData;
}
