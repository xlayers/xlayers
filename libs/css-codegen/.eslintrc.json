{
  "extends": "../../.eslintrc.json",
  "ignorePatterns": ["!**/*"],
  "overrides": [
    {
      "files": ["*.ts"],
      "parserOptions": {
        "project": [
          "libs/css-codegen/tsconfig.lib.json",
          "libs/css-codegen/tsconfig.lib.prod.json",
          "libs/css-codegen/tsconfig.spec.json"
        ],
        "createDefaultProgram": true
      },
      "rules": {
        "@typescript-eslint/consistent-type-definitions": "error",
        "@typescript-eslint/dot-notation": "off",
        "@typescript-eslint/explicit-member-accessibility": [
          "off",
          {
            "accessibility": "explicit"
          }
        ],
        "no-shadow": "off",
        "@typescript-eslint/ban-types": "off"
      }
    },
    {
      "files": ["*.html"],
      "rules": {}
    }
  ]
}
