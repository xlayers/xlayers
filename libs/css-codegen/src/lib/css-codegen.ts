export interface CssCodeGenOptions {
  generateClassName?: boolean;
  cssPrefix?: string;
  componentDir?: string;
  force?: boolean;
}

export interface RuleSet {
  [key: string]: string;
}

export interface CssCodeGenContext {
  rules?: RuleSet;
  className?: string;
  pseudoElements?: { [key: string]: RuleSet };
}
