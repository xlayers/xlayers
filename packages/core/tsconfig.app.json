{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "../../out-tsc/app",
    "module": "es2015",
    "types": ["node", "sketchapp"]
  },
  "exclude": [
    "test.ts",
    "**/*.spec.ts"
  ]
}
