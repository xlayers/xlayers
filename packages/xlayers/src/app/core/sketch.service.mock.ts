import { getFlatLayerMock } from '@app/editor/preview/viewer/sketch-layer/sketch-layer.component.mock';

export function getSketchDataMock() {
  return {
    previews: [{}],
    pages: [
      getFlatLayerMock(9)
    ]
  } as SketchMSData;
}
