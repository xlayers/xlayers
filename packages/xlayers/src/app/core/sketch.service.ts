import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { InformUser, UiState, CurrentFile } from '@app/core/state';
import { environment } from '@env/environment';
import { Store } from '@ngxs/store';
import {
  SketchStyleParserService,
  SupportScore
} from '@xlayers/sketchapp-parser';

export interface SketchMSData {
  pages: SketchMSPage[];
  previews: SketchMSPreview[];
  document: SketchMSDocumentData;
  user: SketchMSUserData;
  meta: SketchMSMetadata;
  resources?: {
    images: {
      [id: string]: ResourceImageData;
    };
  };
}

export interface ResourceImageData {
  source: string;
  image: HTMLImageElement;
}

@Injectable({
  providedIn: 'root'
})
export class SketchService {
  _data: SketchMSData;

  constructor(
    private sketchStyleParser: SketchStyleParserService,
    private http: HttpClient,
    private store: Store
  ) {
    this.store.select(UiState.currentFile).subscribe(currentFile => {
      this._data = currentFile as SketchMSData;
    });
  }

  async process(file: File) {
    const data = await this.sketch2Json(file) as SketchMSData;

    if (this.sketchStyleParser.visit(data as SketchMSData) === SupportScore.LEGACY) {
      this.store.dispatch(
        new InformUser(
          'The design was created using a legacy version of SketchApp, so the result may not be accurate.'
        )
      );
    }
    return data;
  }

  async readZipEntries(file: Blob) {
    return new Promise<any[]>((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = readerEvent => {
        const data = (readerEvent.target as FileReader).result;
        window['JSZip']
          .loadAsync(data)
          .then((zip: any[]) => {
            const zips = [];
            zip.forEach((relativePath: string, zipEntry) => {
              zips.push({ relativePath, zipEntry });
            });
            resolve(zips);
          })
          .catch((e: Error) => {
            reject(e);
          });
      };
      reader.onerror = e => {
        reject(e);
      };
      try {
        reader.readAsArrayBuffer(file);
      } catch (error) {
        reject(error);
      }
    });
  }

  async computeImage(source: string, filepath: string) {
    return new Promise<HTMLImageElement>((resolve, reject) => {
      const image = new Image();
      image.onload = _imageLoadEvent => {
        resolve(image);
      };
      image.onerror = _error => {
        reject(`Could not load a Sketch preview "${filepath}"`);
      };
      image.src = source;
    });
  }

  private async buildImage(
    content: string,
    relativePath: string
  ): Promise<ResourceImageData> {
    const source = `data:image/png;base64,${content}`;
    return {
      source,
      image: await this.computeImage(source, relativePath)
    };
  }

  async sketch2Json(file: Blob) {
    const _data: SketchMSData = {
      pages: [],
      previews: [],
      document: {} as any,
      user: {},
      meta: {} as any,
      resources: {
        images: {}
      }
    };

    const zips = await this.readZipEntries(file);

    await Promise.all(
      zips.map(async ({ relativePath, zipEntry }) => {
        if (
          relativePath === 'previews/preview.png' ||
          relativePath.startsWith('images/')
        ) {
          const content = await zipEntry.async('base64');
          const imageData = await this.buildImage(content, relativePath);

          if (relativePath === 'previews/preview.png') {
            // this is a preview, so add it to the previews array
            _data.previews.push({
              source: imageData.source,
              width: imageData.image.width,
              height: imageData.image.height
            });
          } else {
            // this is a resource image, add it to the resource factory
            _data.resources.images[relativePath] = imageData;
          }
        } else if (relativePath.startsWith('pages/')) {
          const content = await zipEntry.async('string');

          try {
            const page = JSON.parse(content) as SketchMSPage;
            _data.pages.push(page);
          } catch (e) {
            throw new Error(`Could not load page "${relativePath}"`);
          }
        } else if (relativePath.endsWith('.pdf')) {
          // text-previews/text-previews.pdf
          // removed because of: https://github.com/xlayers/xlayers/issues/200
        } else {
          // document.json
          // user.json
          // meta.json
          const content = await zipEntry.async('string');
          _data[relativePath.replace('.json', '')] = JSON.parse(content);
        }
      })
    );

    return _data;
  }

  getPages(): SketchMSPage[] {
    return this._data.pages;
  }

  getDemoFiles() {
    return environment.demoFiles
      .filter(meta => !meta.disabled)
      .sort((m1, m2) => m2.value - m1.value);
  }

  getSketchDemoFile(filename: string) {
    const repoUrl = `${window.location.origin ||
      environment.baseUrl}/assets/demos/sketchapp/`;
    return this.http.get(`${repoUrl}${filename}.sketch`, {
      responseType: 'blob'
    });
  }

  getImageDataFromRef(ref: string) {
    return this._data.resources.images[ref];
  }
}
