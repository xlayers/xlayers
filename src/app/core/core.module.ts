import { ScrollingModule } from '@angular/cdk/scrolling';
import { CommonModule } from '@angular/common';
import { NgModule, ErrorHandler } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { MatButtonModule } from '@angular/material/button';
import { MatButtonToggleModule } from '@angular/material/button-toggle';
import { MatCardModule } from '@angular/material/card';
import { MatExpansionModule } from '@angular/material/expansion';
import { MatIconModule } from '@angular/material/icon';
import { MatInputModule } from '@angular/material/input';
import { MatListModule } from '@angular/material/list';
import { MatMenuModule } from '@angular/material/menu';
import { MatSelectModule } from '@angular/material/select';
import { MatSidenavModule } from '@angular/material/sidenav';
import { MatSlideToggleModule } from '@angular/material/slide-toggle';
import { MatSliderModule } from '@angular/material/slider';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { MatTabsModule } from '@angular/material/tabs';
import { MatToolbarModule } from '@angular/material/toolbar';
import { MatTooltipModule } from '@angular/material/tooltip';
import { MatTreeModule } from '@angular/material/tree';
import { DragDropModule } from '@angular/cdk/drag-drop';
import { ColorSketchModule } from 'ngx-color/sketch';

const MatModules = [
  MatSliderModule,
  MatTreeModule,
  MatSnackBarModule,
  MatExpansionModule,
  MatSlideToggleModule,
  MatListModule,
  MatSidenavModule,
  MatIconModule,
  MatButtonModule,
  MatToolbarModule,
  MatInputModule,
  ScrollingModule,
  MatCardModule,
  MatTabsModule,
  MatButtonToggleModule,
  MatMenuModule,
  MatTooltipModule,
  MatSelectModule,
  DragDropModule
];

const ExtraModules = [FormsModule, ColorSketchModule];
@NgModule({
  imports: [...MatModules, ...ExtraModules],
  exports: [CommonModule, ...MatModules, ...ExtraModules],
})
export class CoreModule { }
