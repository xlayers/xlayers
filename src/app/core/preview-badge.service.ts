import { Inject, Injectable } from '@angular/core';
import { WINDOW } from '@app/core/window.service';

@Injectable({
  providedIn: 'root'
})
export class PreviewBadgeService {
  badge = 'BETA';

  constructor(@Inject(WINDOW) private readonly window: Window) {}

  computeBadge() {
    try {
      const hostname = this.window.location.hostname;
      if (
        hostname === 'localhost' ||
        hostname === '127.0.0.1' ||
        hostname === '0.0.0.0'
      ) {
        return 'LOCAL PREVIEW';
      } else if (hostname.startsWith('next.')) {
        return 'MASTER PREVIEW';
      } else if (hostname.includes('netlify')) {
        return 'PR PREVIEW';
      }

      return this.badge;
    } catch (errror) {
      return this.badge;
    }
  }
}
