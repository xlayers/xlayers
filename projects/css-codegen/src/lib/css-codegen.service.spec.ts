import { TestBed } from '@angular/core/testing';
import { CssCodeGenService } from './css-codegen.service';
import { VERSION_LIST, SKETCH_PATH, loadSketch } from '../../../../tests/test-utils';
import { readdirSync } from 'fs';


describe('CssCodeGenService', () => {
    let service: CssCodeGenService;
    let sketch: SketchMSData;
    beforeEach(async () => {
        TestBed.configureTestingModule({
            providers: [CssCodeGenService]
        });

        service = TestBed.get(CssCodeGenService);
        const version = VERSION_LIST[0];
        const fileNames = readdirSync(`${SKETCH_PATH}/${version}`);

        const fileName = fileNames[1];
        const data = await loadSketch(version, fileName);
        data.pages.forEach(page => {
            service.compute(page, data, {
                generateClassName: false
            });

            sketch = data;
        });
    });

    it('should create', () => {
        expect(service).toBeTruthy();
    });

    it('should provide the context', () => {
        const data = sketch.pages[0].layers[0];
        expect(service.context(data)).toEqual({
            'rules':
            {
                'display': 'block',
                'height': '36px',
                'left': '0px',
                'position': 'absolute',
                'top': '0px',
                'visibility': 'visible',
                'width': '94px'
            }
        });
    });

    it('should provide the context', () => {
        const data = sketch.pages[0].layers[0];
        expect(service.aggregate(data)).toMatchSnapshot();
    });

});
