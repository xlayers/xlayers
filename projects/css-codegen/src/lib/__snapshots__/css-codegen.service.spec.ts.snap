// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssCodeGenService should provide the context 1`] = `
Array [
  Object {
    "kind": "css",
    "language": "css",
    "uri": "components/button.css",
    "value": ":host {
  display: block;
  position: relative;
}
 
.undefined {
  background-color: rgba(0,150,136,1.00);
  box-shadow: 0px 8px 8px 0px rgba(0,0,0,0.24),0px 0px 8px 0px rgba(0,0,0,0.12),0px 2px 2px 0px rgba(0,0,0,0.24),0px 0px 2px 0px rgba(0,0,0,0.12);
  height: 36px;
  left: 0px;
  top: 0px;
  width: 94px;
}

.undefined {
  color: rgba(255,255,255,1.00);
  font-family: 'Roboto-Medium', 'Roboto', 'sans-serif';
  font-size: 14px;
  height: 16px;
  left: 16.5px;
  top: 10px;
  width: 61px;
}

.undefined, .undefined {
  display: block;
  position: absolute;
  visibility: visible;
}
",
  },
]
`;
